src/lib/storage/dictionary_segment.hpp

```diff
@@ -25,55 +30,128 @@ class DictionarySegment : public BaseSegment {
 
   // return the number of unique_values (dictionary entries)
-  size_t unique_values_count() const;
+  size_t unique_values_count() const { return _dictionary->size(); }
 
   // return the number of entries
-  size_t size() const override;
+  size_t size() const override { return _attribute_vector->size(); }
 
   // returns the calculated memory usage
-  size_t estimate_memory_usage() const final;
+  size_t estimate_memory_usage() const final {
+    const size_t dictionary_size = _dictionary->capacity() * sizeof(T);
+    return dictionary_size + _attribute_vector->estimate_memory_usage();
+  }
 
  protected:
   std::shared_ptr<std::vector<T>> _dictionary;
   std::shared_ptr<BaseAttributeVector> _attribute_vector;
+
+  void _compress_values(const std::vector<T>& column_values) {
+    std::vector<uint32_t> lookup_indices(column_values.size());
+    // initialized with many falses
+    std::vector<bool> same_as_previous_element(column_values.size());
+
+    std::iota(lookup_indices.begin(), lookup_indices.end(), 0);
+    std::sort(lookup_indices.begin(), lookup_indices.end(),
+              [&column_values](const uint32_t index_1, const uint32_t index_2) {
+                return column_values[index_1] < column_values[index_2];
+              });
+
+    uint32_t previous_uncompressed_index = lookup_indices[0];
+    uint32_t num_unique = 1;
+    for (uint32_t position = 1; position < column_values.size(); ++position) {
+      uint32_t uncompressed_index = lookup_indices[position];
+      if (column_values[previous_uncompressed_index] == column_values[uncompressed_index]) {
+        same_as_previous_element[position] = true;
+      } else {
+        ++num_unique;
+      }
+      previous_uncompressed_index = uncompressed_index;
+    }
+
+    if (num_unique <= static_cast<uint32_t>(std::numeric_limits<uint8_t>::max()) + 1) {
+      _build_dictionary_and_attributes<uint8_t>(column_values, lookup_indices, same_as_previous_element, num_unique);
+    } else if (num_unique <= static_cast<uint32_t>(std::numeric_limits<uint16_t>::max()) + 1) {
+      _build_dictionary_and_attributes<uint16_t>(column_values, lookup_indices, same_as_previous_element, num_unique);
+    } else {
+      _build_dictionary_and_attributes<uint32_t>(column_values, lookup_indices, same_as_previous_element, num_unique);
+    }
+  }
+
+  template <typename IndexType>
+  void _build_dictionary_and_attributes(const std::vector<T>& values, const std::vector<uint32_t>& indices,
+                                        const std::vector<bool>& same_as_before, const uint32_t num_unique) {
+    std::vector<IndexType> attributes(values.size());
+    _dictionary = std::make_shared<std::vector<T>>();
+    _dictionary->reserve(num_unique);
+
+    for (uint32_t position = 0; position < values.size(); position++) {
+      uint32_t uncompressed_index = indices[position];
+      if (!same_as_before[position]) {
```

Comment:
Since the code looks like you already tried to squeeze out the last bit of performance: I feel like you might even be able to drop the `same_as_before` vector by using three linear iterations, one counting the unique values (maybe even remove that and just rely on amortized push_back runtime? Or just reserve for as many values are in the value segment and shrink_to_fit it afterwards?), then making the dictionary and reserve enough space, then build the dictionary, and then go through it and build up the attributes. Depending on how many duplicate values there are (and thus how often you'd get cache misses), I think there is a chance of that being faster -- obviously, you'd save some memory. Have you tried that / intentionally decided not to do that?

Reply:
I was interested in the performance, so I tried around a bit. Sharing in case you care. With your code, the test suite passed in 7ms on my machine. If I change the `_compress_chunk` method to the following code, it the suite runs in about 3ms:
```c++
  void _compress_values(const std::vector<T>& column_values) {
    const size_t value_count = column_values.size();

    std::vector<uint32_t> lookup_indices(column_values.size());
    std::iota(lookup_indices.begin(), lookup_indices.end(), 0);
    std::sort(lookup_indices.begin(), lookup_indices.end(),
              [&column_values](const uint32_t& index_1, const uint32_t& index_2) {
                return column_values[index_1] < column_values[index_2];
              });

    uint32_t previous_uncompressed_index = lookup_indices[0];
    uint32_t num_unique = 1;
    // TODO: If there are values that are duplicated more often, this can be improved by binary searching for the end of
    // the repitition
    for (uint32_t position = 1; position < value_count; ++position) {
      uint32_t uncompressed_index = lookup_indices[position];
      if (column_values[previous_uncompressed_index] != column_values[uncompressed_index]) {
        ++num_unique;
      }
      previous_uncompressed_index = uncompressed_index;
    }

    if (num_unique <= static_cast<uint32_t>(std::numeric_limits<uint8_t>::max()) + 1) {
      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint8_t>>(column_values.size());
    } else if (num_unique <= static_cast<uint32_t>(std::numeric_limits<uint16_t>::max()) + 1) {
      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint16_t>>(column_values.size());
    } else {
      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint32_t>>(column_values.size());
    }

    previous_uncompressed_index = lookup_indices[0];
    _dictionary = std::make_shared<std::vector<T>>(num_unique);
    (*_dictionary)[0] = column_values[previous_uncompressed_index];
    uint32_t dictionary_index = 1;

    for (uint32_t position = 0; position < value_count; position++) {
      uint32_t uncompressed_index = lookup_indices[position];
      if (column_values[previous_uncompressed_index] != column_values[uncompressed_index]) {
        (*_dictionary)[dictionary_index++] = column_values[uncompressed_index];
      }
      _attribute_vector->set(uncompressed_index, static_cast<ValueID>(dictionary_index - 1));

      previous_uncompressed_index = uncompressed_index;
    }
  }
```